### PR TITLE
User API endpoints (code-first approach) 1st attempt

### PR DESCRIPTION
**Pull request by karpov-kir**
This is my first attempt to implement the user API endpoints assignment - https://www.essentialist.dev/products/the-software-essentialist/categories/2153149734/posts/2168948146.

I used GitHub Copilot X to build the solution for me. I changed the code just slightly (literally changed a few words). The final solution looks surpassingly fine for the code-first approach. I think I could tweak the solution even more using Copilot, but since this is a code-first approach there is no need to polish it.

Here is a full command list I send to Copilot with no edits:

<details>
<summary>GitHub Copilot X commands</summary>

- Could you help me build an API that exposes create user, edit user, get user by email endpoints. The solutions should use Express, Typescript, Prisma, Sqlite3.

- Could your rewrite your solution to use MVC pattern?

- Could you move logic to services?

- Could you create Prisma models for the solution above?

- How to use the model you've created?

- How to define Sqlite datasource in my schema.prisma?

- How to define a generator in my shema.prisma for the solution you've created?

- Could you split name into first name and last name in your solution?

- Could you add createdAt and updatedAt to your solution?

- createUser does not work, it says that property 'id' is missing in the generated type.

- Could you omit id from createUser of UserService in your solution?

- How to import Prisma.UserCreateInput?

- How to create tables in the database?

- Could you use id of the user in PUT endpoint?

- Could you user query parameters to get the user's email in the get user by email endpoint?

- Could you add email validation to the solution?

- Could you do the same for updateUser?

- Could you handle invalid email errors gracefuly?

- Could you handle duplicate email error gracefully?

- It does not work with typescript, could you update your implementation?

- Could you handle duplicate email error gracefuly in typescript?

- I get a type error "Property 'includes' does not exist on type '{}'.", how to fix that?
</details>

API:

- `POST to /users`: Creates a new user
- `PUT to /users/:userId`: Edits a user
- `GET /users?email=someuser@gmail.com`: Fetches a user by their email

How to start locally:

- `cd 0_From_Code_First_To_First_Principles\1_code_first\assignment\src`
- `npm ci`
- `cd src`
- `npx prisma generate`
- `npx prisma migrate deploy`
- `npm run start`

### CODE REVIEW

File: 0_From_Code_First_To_First_Principles/1_code_first/src/index.ts [hunk 1]
```diff
@@ -13,9 +13,20 @@ class UserService {
...
 
-    return await this.prisma.user.create({
-      data: user,
-    });
+    try {
+      return await this.prisma.user.create({
+        data: user,
+      });
+    } catch (error) {
+      const prismaError = error as Prisma.PrismaClientKnownRequestError;
```
[the remark this answers is not in the document]
Comment (karpov-kir) [reply 0_From_Code_First_To_First_Principles/1_code_first/src/index.ts:21]: Yeah, I was surprised that it actually suggested adjusting the imports and found the right one, and even showed how to cast the error correctly. I am now not that skeptical about https://github.com/smol-ai/developer. 😄

**Comment by admiralrohan** [comment]
@karpov-kir What do you do with these copilot comments afterwards?

**Comment by karpov-kir** [comment]
@admiralrohan, hey! 

Within this task, I evaluated how far I could go just using GH Copilot. As I mentioned, it managed to completely generate the complete solution with satisfies code-first requirements.

I will just merge this code a little bit later and call this assignment finished. But the key takeaway is that it was extremely helpful to answer some of my particular questions e.g. how to use Prisma. Therefore, In future assignments (e.g. best practices first) I'm going to write code on my own but use GH Copilot as a consultant to answer some of my questions.

**Comment by karpov-kir** [comment]
I merged the PR since I received and addressed the feedback [on Discord](https://discord.com/channels/997940575590416435/1121693001664708638/1121693001664708638), so this iteration of the assignment can be considered as finished.

**Pull request merged**